pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,654 @@
+					continue
+				}
+
+				// if the destination tag already has this manifest, do nothing
+				var mustCopyLayers bool
+				if o.Force {
+					mustCopyLayers = true
+				} else {
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
+						default:
+							digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("the manifest type %T is not supported", srcManifest)})
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							// tagging within the same registry is a no-op
+							if src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String() {
+								continue
+							}
+
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+
+							var options []distribution.BlobCreateOption
+							if !o.SkipMount {
+								options = append(options, client.WithMountFrom(blobSource), WithDescriptor(blob))
+							}
+							w, err := toBlobs.Create(ctx, options...)
+							// no-op
+							if err == ErrAlreadyExists {
+								glog.V(5).Infof("Blob already exists %#v", blob)
+								continue
+							}
+							// mount successful
+							if ebm, ok := err.(distribution.ErrBlobMounted); ok {
+								glog.V(5).Infof("Blob mounted %#v", blob)
+								if ebm.From.Digest() != blob.Digest {
+									digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to push %s: tried to mount blob %s src source and got back a different digest %s", src.ref, blob.Digest, ebm.From.Digest())})
+									break
+								}
+								continue
+							}
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to upload blob %s to %s: %v", blob.Digest, dst.ref, err)})
+								break
+							}
+
+							err = func() error {
+								glog.V(5).Infof("Uploading blob %s", blob.Digest)
+								defer w.Cancel(ctx)
+								r, err := srcBlobs.Open(ctx, blob.Digest)
+								if err != nil {
+									return fmt.Errorf("unable to open source layer %s to copy to %s: %v", blob.Digest, dst.ref, err)
+								}
+								defer r.Close()
+
+								switch dst.t {
```

Comment:
This switch repeated 3 times. Looks like lack of a new type.

Reply:
As this evolves I think this would move to more of a work queue model - the processor queues up images to move, which are then tried and retried until they succeed.  If we get more types I think this should definitely be abstracted, but wanted 2 examples before it became necessary.